mypy/semanal_typeddict.py

```diff
@@ -67,16 +67,24 @@ def analyze_typeddict_classdef(self, defn: ClassDef) -> Tuple[bool, Optional[Typ
                 defn.analyzed.line = defn.line
                 defn.analyzed.column = defn.column
                 return True, info
+
             # Extending/merging existing TypedDicts
-            if any(not isinstance(expr, RefExpr) or
-                   expr.fullname not in TPDICT_NAMES and
-                   not self.is_typeddict(expr) for expr in defn.base_type_exprs):
-                self.fail("All bases of a new TypedDict must be TypedDict types", defn)
-            typeddict_bases = list(filter(self.is_typeddict, defn.base_type_exprs))
+            typeddict_bases = []
+            typeddict_bases_set = set()
+            for expr in defn.base_type_exprs:
+                if (not isinstance(expr, RefExpr) or
```

Comment:
I feel like there's two many nots here for this code to be readable. It looks like there are three cases: (1) the base class is TypedDict of some sort (expr.fullname in TPDICT_NAMES); (2) the base class is another TypedDict; (3) anything else, which produces an error. The code would be more readable if we covered those three cases in order, without any nots.